syntax = "proto3";

package todos;

import "google/protobuf/struct.proto";

service Todos {
  rpc ListTodos (ListRequest) returns (ListResponse) {}
  rpc CreateTodo (CreateRequest) returns (ItemResponse) {}
  rpc GetTodo (GetRequest) returns (ItemResponse) {}
  rpc UpdateTodo (UpdateRequest) returns (ItemResponse) {}
  rpc DeleteTodo (DeleteRequest) returns (ResultResponse) {}
}

// *** Generic RPC Messages <Start> version 1 ***
message GetRequest {
  int32 id = 1;
}
message ListRequest {
  google.protobuf.Struct filters = 1;
  int32 offset = 2;
  int32 limit = 3;
  repeated string ordering = 4;
}
message CreateRequest {
  google.protobuf.Struct data = 1;
}
message UpdateRequest {
  int32 id = 1;
  google.protobuf.Struct data = 2;
}
message DeleteRequest {
  int32 id = 1;
}
message ItemResponse {
  google.protobuf.Struct data = 1;
}
message ListResponse {
  repeated google.protobuf.Struct data = 1;
  int32 count = 2;
}
message ResultResponse {
  bool result = 1;
}
// *** Generic RPC Messages <End> ***
